File: LeetCode/Pascal's Triangle II/cpp/pascal_triangle.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    public:
        int getNum(int row, int col) {
            if (row == 0 || col == 0 || row == col)
            return 1;

            return getNum(row - 1, col - 1) + getNum(row - 1, col);
        }

        vector<int> getRow(int rowIndex) {
            vector<int> ans;

            for (int i = 0; i <= rowIndex; i++)
            ans.push_back(getNum(rowIndex, i));

            return ans;
        }

        unordered_map<size_t, int> cache;

        // use a better hashing function like `boost::hash_combine` in the real world.
        int key(int i, int j) const {
            size_t hash_i = hash<int>{}(i), hash_j = hash<int>{}(j);
            int hashed = (int)(hash_i ^ (hash_i >> 32));
            return (hashed << 5) - 1 + (int)(hash_j ^ (hash_j >> 32));
        }

        int getNumDP(int row, int col) {
            auto rowCol = key(row, col);

            if (cache.count(rowCol) > 0)
            return cache[rowCol];

            if (row == 0 || col == 0 || row == col)
            return (cache[rowCol] = 1);

            return (cache[rowCol] = getNumDP(row - 1, col - 1) + getNumDP(row - 1, col));
        }

        vector<int> getRowDP(int rowIndex) {
            vector<int> ans;

            for (int i = 0; i <= rowIndex; i++)
            ans.push_back(getNum(rowIndex, i));

            return ans;
        }

          vector<int> getRowDPMemoryEfficient(int rowIndex) {
            vector<int> curr, prev = {1};

            for (int i = 1; i <= rowIndex; i++) {
            curr.assign(i + 1, 1);

            for (int j = 1; j < i; j++)
                curr[j] = prev[j - 1] + prev[j];

            prev = move(curr);  // This is O(1)
            }

            return prev;
        }

};


int main() {

    Solution solution;

    vector<int> res = solution.getRowDP(5);
    
    for(int ele: res) {
        cout << ele;
    }

    return 0;
}

File: LeetCode/Pascal's Triangle II/java/PascalTriangle.java
import java.util.*;

public class PascalTriangle {

    public static int calculate(int row, int col) {

        if(row == 0 || col == 0 || row == col) {
            return 1;
        }

        return calculate(row-1, col) + calculate(row-1, col-1);
    }

    public static List<Integer> getRow(int rowIndex) {
        ArrayList<Integer> res = new ArrayList<>();
        for(int i = 0; i < rowIndex+1; i++) {
            res.add(calculate(rowIndex, i));
        }
        return res;
    }

    // Dynamic Programming
    private static class RowCol {
        private int row, col;

        public RowCol(int row, int col) {
            this.row = row;
            this.col = col;
        }

        @Override
        public int hashCode() {
            int result = (int) (row ^ (row >>> 32));
            return (result << 5) - 1 + (int) (col ^ (col >>> 32)); // 31 * result == (result << 5) - 1
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null) return false;
            if (this.getClass() != o.getClass()) return false;

            RowCol rowCol = (RowCol) o;
            return row == rowCol.row && col == rowCol.col;
        }
    }

    private static Map<RowCol, Integer> cache = new HashMap<>();

    private static int getNum(int row, int col) {
        RowCol rowCol = new RowCol(row, col);

        if (cache.containsKey(rowCol)) {
            return cache.get(rowCol);
        }

        int computedVal =
                (row == 0 || col == 0 || row == col) ? 1 : getNum(row - 1, col - 1) + getNum(row - 1, col);

        cache.put(rowCol, computedVal);

        return computedVal;
    }

    public static List<Integer> getRowDP(int rowIndex) {
        List<Integer> ans = new ArrayList<>();

        for (int i = 0; i <= rowIndex; i++) {
            ans.add(getNum(rowIndex, i));
        }

        return ans;
    }

    public List<Integer> getRowDPMemoryEfficient(int rowIndex) {
        List<Integer> row =
                new ArrayList<>(rowIndex + 1) {
                    {
                        add(1);
                    }
                };

        for (int i = 0; i < rowIndex; i++) {
            for (int j = i; j > 0; j--) {
                row.set(j, row.get(j) + row.get(j - 1));
            }
            row.add(1);
        }

        return row;
    }

    public List<Integer> getRowMath(int rowIndex) {
        List<Integer> row =
                new ArrayList<>(rowIndex + 1) {
                    {
                        add(1);
                    }
                };

        for (int i = 0; i < rowIndex; i++) {
            for (int j = i; j > 0; j--) {
                row.set(j, row.get(j) + row.get(j - 1));
            }
            row.add(1);
        }

        return row;
    }

    public static void main(String[] args) {
        int val = 5;
        System.out.println(Arrays.toString(getRow(5).toArray()));
        System.out.println(Arrays.toString(getRowDP(5).toArray()));
    }
}
